lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -117,6 +117,136 @@ static T &operator<<(T &os, const SignalMapping &mapping) {
+                        prefix.consume_front("|"));
+  fields.append("isSubCircuit", BoolAttr::get(context, isSubCircuit));
+
+  auto dontTouch = [&](StringRef targetStr) {
+    NamedAttrList anno;
+    anno.append("class", StringAttr::get(context, dontTouchAnnoClass));
+    anno.append("target", StringAttr::get(context, targetStr));
+    state.addToWorklistFn(DictionaryAttr::get(context, anno));
+  };
+
+  StringSet<> modules;
+  auto buildSourceOrSinkAnno = [&](Attribute attr,
+                                   bool isSource) -> DictionaryAttr {
+    auto dict = attr.dyn_cast<DictionaryAttr>();
+    if (!dict)
+      return {};
+    NamedAttrList fields;
+    fields.append("class",
+                  StringAttr::get(context, signalDriverTargetAnnoClass));
+    StringAttr targetAttr = tryGetAs<StringAttr>(
+        dict, dict, isSubCircuit ? "_2" : "_1", loc, signalDriverAnnoClass);
+    StringAttr peer = tryGetAs<StringAttr>(
+        dict, dict, isSubCircuit ? "_1" : "_2", loc, signalDriverAnnoClass);
+    if (!targetAttr)
+      return {};
+
+    auto target = tokenizePath(targetAttr);
+    if (!target)
+      return {};
+    target->component.clear();
+    target->name = target->name.drop_front(target->name.size());
+    modules.insert(target->str());
+
+    auto targetId = state.newID();
+    fields.append("dir",
+                  StringAttr::get(context, isSource ? "source" : "sink"));
+    fields.append("id", id);
+    fields.append("peer", peer);
+    fields.append("side",
+                  StringAttr::get(context, isSubCircuit ? "local" : "remote"));
+    fields.append("target", targetAttr);
+    fields.append("targetId", targetId);
+
+    // TODO: Always applying DontTouchAnnotation to a target is too restrictive.
+    // This should only be necessary if this is a source (which will produce a
+    // force) and we are in the circuit where the force will be applied.  Relax
+    // this restriction in the future.
+    dontTouch(targetAttr.getValue());
+
+    return DictionaryAttr::get(context, fields);
+  };
+
+  // Add any sinks or sources to the worklist for later processing.
+  for (auto attr : sourcesAttr) {
+    auto sourceAnno = buildSourceOrSinkAnno(attr, true);
+    if (!sourceAnno)
+      return failure();
+    state.addToWorklistFn(sourceAnno);
+  }
+  for (auto attr : sinksAttr) {
+    auto sourceAnno = buildSourceOrSinkAnno(attr, false);
+    if (!sourceAnno)
+      return failure();
+    state.addToWorklistFn(sourceAnno);
+  }
+
+  AnnotationSet annotations(state.circuit);
+  annotations.addAnnotations({DictionaryAttr::get(context, fields)});
+  annotations.applyToOperation(state.circuit);
+
+  for (auto &module : modules) {
+    NamedAttrList fields;
+    fields.append("class",
+                  StringAttr::get(context, signalDriverModuleAnnoClass));
+    fields.append("id", id);
+    fields.append("target", StringAttr::get(context, module.getKey()));
```

Comment:
Aw, lame that StringSet exposes this detail to its users (such that you have to use `getKey()`).

Reply:
Yeah, this felt like a sharp edge...